File: app/src/main/java/uk/co/transferx/app/ui/recoverpass/fragment/RecoverPasswordFragment.java
package uk.co.transferx.app.ui.recoverpass.fragment;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputEditText;
import android.support.v4.content.ContextCompat;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import javax.inject.Inject;

import uk.co.transferx.app.ui.base.BaseFragment;
import uk.co.transferx.app.R;
import uk.co.transferx.app.TransferXApplication;
import uk.co.transferx.app.ui.recoverpass.RecoverPasswordActivity;
import uk.co.transferx.app.ui.recoverpass.presenter.RecoverPasswordPresenter;

/**
 * Created by sergey on 23.11.17.
 */

public class RecoverPasswordFragment extends BaseFragment implements RecoverPasswordPresenter.RecoverPasswordUI {

    @Inject
    RecoverPasswordPresenter presenter;

    private TextInputEditText textInputEditText;
    private TextView label, firstError;
    private Button sendEmail;

    @Override
    public String tagName() {
        return RecoverPasswordFragment.class.getSimpleName();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((TransferXApplication) (getActivity().getApplication())).getAppComponent().inject(this);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.recover_password_fragment_layout, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        textInputEditText = view.findViewById(R.id.mailInput);
        label = view.findViewById(R.id.mailLabel);
        firstError = view.findViewById(R.id.mailError);
        textInputEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
               // setStatusOfError(textInputEditText, label, R.color.black);
               // firstError.setVisibility(View.GONE);
                presenter.validateEmail(s.toString());
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
        sendEmail = view.findViewById(R.id.send_mail_button);
        sendEmail.setOnClickListener(v -> presenter.sendEmail());
        view.findViewById(R.id.back_button).setOnClickListener(v -> presenter.goBack());
    }

    @Override
    public void unlockButton() {
        sendEmail.setEnabled(true);
        sendEmail.setBackground(ContextCompat.getDrawable(getContext(), R.drawable.oval_button_black));
        sendEmail.setTextColor(ContextCompat.getColor(getContext(), R.color.white));
      //  setStatusOfError(textInputEditText, label, R.color.red);
      //  firstError.setText(R.string.email_error);
      //  firstError.setVisibility(View.VISIBLE);
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.attachUI(this);
    }

    @Override
    public void successGoBack() {
        hideKeyboard(textInputEditText);
        ((RecoverPasswordActivity) getActivity()).goSuccess();
    }

    @Override
    public void onPause() {
        presenter.detachUI();
        super.onPause();
    }

    @Override
    public void lockButton() {
        sendEmail.setEnabled(false);
        sendEmail.setBackground(ContextCompat.getDrawable(getContext(), R.drawable.oval_button_gray));
    }

    @Override
    public void goBackToMain() {
        getActivity().finish();
    }

    @Override
    public void error() {
        Toast.makeText(getContext(), "Error occurred", Toast.LENGTH_SHORT).show();
    }

    @Override
    public void goToWelcome() {
        // no op
    }
}

File: app/src/main/java/uk/co/transferx/app/ui/recoverpass/fragment/RecoverPasswordFragment.kt
package uk.co.transferx.app.ui.recoverpass.fragment

import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.design.widget.TextInputEditText
import android.support.v4.content.ContextCompat
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import kotlinx.android.synthetic.main.recover_password_fragment_layout.*

import javax.inject.Inject

import uk.co.transferx.app.ui.base.BaseFragment
import uk.co.transferx.app.R
import uk.co.transferx.app.TransferXApplication
import uk.co.transferx.app.ui.recoverpass.RecoverPasswordActivity
import uk.co.transferx.app.ui.recoverpass.RecoverPasswordContract
import uk.co.transferx.app.ui.recoverpass.presenter.RecoverPasswordPresenter

/**
 * Created by sergey on 23.11.17.
 */

class RecoverPasswordFragment : BaseFragment(), RecoverPasswordContract.View {
    @Inject
    lateinit var presenter: RecoverPasswordPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (activity?.application as TransferXApplication).appComponent.inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.recover_password_fragment_layout, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        sendEmail?.setOnClickListener { presenter.sendEmail() }
        buttonBack.setOnClickListener { presenter.goBack() }
    }

    override fun unlockButton() {
        sendEmail?.isEnabled = true
        sendEmail?.background = ContextCompat.getDrawable(context!!, R.drawable.btn_style_bright)
    }

    override fun onResume() {
        super.onResume()

        presenter.attachUI(this)
        recoverMailInputText.addTextChangedListener(recoverEmailTextWatcher)
    }

    override fun onPause() {
        super.onPause()

        presenter.detachUI()
        recoverMailInputText.removeTextChangedListener(recoverEmailTextWatcher)
    }

    override fun goToSuccessScreen() {
        hideKeyboard(recoverMailInputText)
        (activity as RecoverPasswordActivity).goSuccess()
    }

    override fun lockButton() {
        sendEmail?.isEnabled = false
        sendEmail?.background = ContextCompat.getDrawable(context!!, R.drawable.btn_disabled)
    }

    override fun goBackToMain() {
        activity?.finish()
    }

    override fun error() {
        Snackbar.make(rootLayout, getString(R.string.connection_error), Snackbar.LENGTH_LONG)
    }

    override fun goToWelcome() {
        // no op
    }

    override fun tagName(): String {
        return RecoverPasswordFragment::class.java.simpleName
    }

    private val recoverEmailTextWatcher = object : TextWatcher {
        override fun afterTextChanged(s: Editable?) {
        }

        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
        }

        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            presenter.validateEmail(s.toString())
        }
    }
}
